packages/web/components/Guild/GuildForm.tsx

```diff
@@ -205,9 +210,104 @@ export const GuildForm: React.FC<Props> = ({
     reset(values);
   }, [workingGuild, guildMetadata, roleOptions, reset]);
 
+  const [logoURI, setLogoURI] = useState<string | undefined>(
+    workingGuild.logo ?? undefined,
+  );
+  const [active, setActive] = useState(false);
+  const [loading, setLoading] = useState(true);
+  const [errored, setErrored] = useState(false);
+  const toast = useToast();
+
+  const onFileChange = useCallback(
+    (file: File | undefined) => {
+      if (!file) return;
+      setLoading(true);
+      setErrored(false);
+      const reader = new FileReader();
+      reader.addEventListener('load', () => {
+        setLogoURI(reader.result as string);
+      });
+      reader.addEventListener('error', ({ target }) => {
+        const { error } = target ?? {};
+        toast({
+          title: 'Image Loading Error',
+          description: `Loading Images Error: “${error?.message}”`,
+          status: 'error',
+          isClosable: true,
+          duration: 10000,
+        });
+        setLoading(false);
+        setErrored(true);
+      });
+      reader.readAsDataURL(file);
+    },
+    [toast],
+  );
+
   return (
     <Box w="100%" maxW="40rem">
       <VStack>
+        <Field label="Logo" error={errors.logoUrl}>
+          <Flex
+            w="10em"
+            h="10em"
+            borderRadius="full"
+            display="inline-flex"
+            overflow="hidden"
+            align="center"
+            justify="center"
+            position="relative"
+            border="2px solid"
+            borderColor={active ? 'blue.400' : 'transparent'}
+          >
+            <Image
+              onLoad={() => setLoading(false)}
+              onError={() => setErrored(true)}
+              display={loading ? 'none' : 'inherit'}
+              src={logoURI}
+              borderRadius="full"
+              objectFit="cover"
+              h="full"
+              w="full"
+            />
+            {loading &&
+              (!logoURI || errored ? (
+                <Image w="5em" mx="2.5em" src={FileOpenIcon} opacity={0.5} />
+              ) : (
+                <Spinner size="xl" color="purple.500" thickness="4px" />
+              ))}
```

Comment:
I close grouping symbols on separate lines if they were opened on separate lines:

```suggestion
              ))
            }
```